crates/prune/src/pruner.rs

```diff
@@ -146,69 +179,82 @@ impl<DB: Database> Pruner<DB> {
-                    ?purpose,
-                    %to_block,
-                    ?prune_mode,
-                    %output.pruned,
-                    "Segment pruning finished"
-                );
-
-                if output.pruned > 0 {
-                    stats.insert(
-                        segment.segment(),
-                        (PruneProgress::from_done(output.done), output.pruned),
+                    let segment_start = Instant::now();
+                    let previous_checkpoint = provider.get_prune_checkpoint(segment.segment())?;
+                    let segment_limiter = segment.new_limiter_from_parent_scope_limiter(&limiter);
+                    let output = segment.prune(
+                        provider,
+                        PruneInput { previous_checkpoint, to_block, limiter: segment_limiter },
+                    )?;
+                    if let Some(checkpoint) = output.checkpoint {
+                        segment.save_checkpoint(
+                            provider,
+                            checkpoint.as_prune_checkpoint(prune_mode),
+                        )?;
+                    }
+                    self.metrics
+                        .get_prune_segment_metrics(segment.segment())
+                        .duration_seconds
+                        .record(segment_start.elapsed());
+
+                    self.metrics
+                        .get_prune_segment_metrics(segment.segment())
+                        .highest_pruned_block
+                        .set(to_block as f64);
+
+                    done = done && output.progress.is_done();
+                    limiter.increment_deleted_entries_count_by(output.pruned);
+
+                    debug!(
+                        target: "pruner",
+                        segment = ?segment.segment(),
+                        ?purpose,
+                        %to_block,
+                        ?prune_mode,
+                        %output.pruned,
+                        "Segment pruning finished"
                     );
+
+                    if output.pruned > 0 {
+                        // sets `is_timed_out`
+                        let _ = limiter.is_limit_reached();
+
+                        stats.insert(segment.segment(), (output.progress, output.pruned));
+                    }
+                } else {
+                    debug!(target: "pruner", segment = ?segment.segment(), ?purpose, "Nothing to prune for the segment");
+                }
+
+                if limiter.is_limit_reached() {
+                    break
                 }
-            } else {
-                debug!(target: "pruner", segment = ?segment.segment(), ?purpose, "Nothing to prune for the segment");
             }
         }
 
-        Ok((stats, delete_limit, PruneProgress::from_done(done)))
+        Ok((
+            stats,
+            limiter.deleted_entries_count(),
+            PruneProgress::new(done, limiter.is_timed_out()),
```

Comment:
I think this API is confusing. Can we just construct the enum manually? Because passing the `timeout: bool` and falling back to the entries limit as a default seems easy to misuse